Style book: create static categories

ramonjd opened the pull request:
Part of:

- https://github.com/WordPress/gutenberg/issues/64707

## What? How?

MVP for https://github.com/WordPress/gutenberg/issues/53431

- Creates a static category map based on list defined in #53431 
- Sorts examples based on static categories and their subcategories
- Introduces simple layout for subcategories (under theme)


## Why?

> [!NOTE]
> This is a foundational PR only. So clearing a path to creating custom tabs and example lists. More design stuff to come in subsequent PRs. 

Below is the theme tab for example. The "Design" tab's examples are now under "Theme" sorted into subcategories.


## Testing Instructions
Open the style book and check for regressions, e.g., 

1. Click on examples. They should open the global styles panel for that block.
2. Navigate between tabs
3. Check that the stylebook can be opened from the "Style" menu item in the site editor "view" mode navigation menu. 

## Screenshots

| Trunk | This PR |
|--------|--------|
| <img width="500" alt="Screenshot 2024-09-20 at 12 52 11 pm" src="https://github.com/user-attachments/assets/61109574-a9f5-4b4d-a1b3-be6fdbad332d"> | <img width="500" alt="Screenshot 2024-09-20 at 12 51 55 pm" src="https://github.com/user-attachments/assets/768a63cb-9caa-44a9-bbc7-263921cfa269"> | 

## Steps after this PR:

- Implement layout/design in #53431 
- Create custom "Color" pattern to display theme's palettes/other colors.
- Create/collate example blocks for the landing category. Include said category. Add click events that open corresponding category tabs.
- Look at how themes can define their own examples for the "Overview" category.
- Ensure registered categories not in the static list are appended to the static list via `getCategories()` for extenders etc. https://github.com/WordPress/gutenberg/pull/65507
- Come up with a way for the UI to display lots of categories.
- Create missing block examples where appropriate.

ramonjd committed:
Initial commit - working out category model

ramonjd committed:
More stuff

ramonjd committed:
Working out the category model

ramonjd committed:
Adding test

ramonjd committed:
Start integration

ramonjd committed:
Groups

ramonjd committed:
Clean up experiments and unused code for MVP

ramonjd committed:
Remove comments

ramonjd committed:
Tinkering around the edges with subcategory reducer

ramonjd committed:
I found the custom link! :)

ramonjd commented on packages/edit-site/src/components/style-book/utils.js:
```diff
@@ -0,0 +1,107 @@
+} from '@wordpress/blocks';
+
+// Get all examples from every registered block.
+export function getExamples() {
```
utils.js is probably not a very intuitive name for this file but what ya gunna do

aaronrobertshaw replied:
> but what ya gunna do

Call it `examples.js`, `get-examples.js` or something?

ramonjd replied:
The file contains category filters. I'm also working on getting the full list of registered categories over in https://github.com/WordPress/gutenberg/pull/65507

So maybe split up the file into `examples.js`, `categories.js` or 🤷🏻 

Naming here could be a future problem when things are fleshed out.

aaronrobertshaw replied:
> Naming here could be a future problem when things are fleshed out.

Agreed.

It might be nice before the current iterations on the Style Book are completed to split it up. Generic util files for these sorts of things tend to get a bit out of hand over time.

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -402,11 +271,45 @@ const Examples = memo(
+							key={ `subcategory-${ subcategory.name }` }
+						>
+							<Composite.GroupLabel className="edit-site-style-book__subcategory-label">
+								{ subcategory.title }
```
What this should look like in the interim is the question.

<img width="772" alt="Screenshot 2024-09-19 at 4 50 07 pm" src="https://github.com/user-attachments/assets/0b17c3b7-9db2-4e8f-a142-baebce711aee">

ramonjd committed:
Tweaking subcategory headings. Guard against empty category examples.

ramonjd committed:
Add default category

Using fixtures for tests.

Using fixtures for tests.

ramonjd commented on packages/edit-site/src/components/style-book/test/utils.js:
```diff
@@ -0,0 +1,41 @@
+			);
+			expect( themeCategoryExamples.name ).toEqual( 'theme' );
+			expect( themeCategoryExamples.subcategories ).toHaveLength(
+				themeCategory.subcategories.length
```
I'm thinking I might use some test fixtures instead of all the core blocks to make things more predictable and "unit"-like.

vcanales commented:
Somewhat related, [this PR](https://github.com/WordPress/gutenberg/pull/64705) intended to iterate over the designs for the new categories.

Tracking some of these efforts here: https://github.com/WordPress/gutenberg/issues/64707

ramonjd commented:
@vcanales Thanks for linking.

Also for the [tracking issue](https://github.com/WordPress/gutenberg/issues/64707) 👍🏻  Sorry, I didn't see that until today. 

Is there anything else e can add to the TODO list in the description of this PR do you think? 

Let's coordinate so we don't double up on work. 

I'm currently looking at catering for other registered categories (aside from those in the static list).

I think @aaronrobertshaw might be working on custom examples, e.g., the theme colors.

ramonjd committed:
update e2e

aaronrobertshaw commented:
> I think @aaronrobertshaw might be working on custom examples, e.g., the theme colors.

I'm currently working through ensuring we have examples for all the core blocks outlined in the style book iteration issue. 

I've gotten a little sidetracked on some blocks that render placeholder states, such as the Avatar block. It might be open for debate as to what state the block should be in for the Style Book but that can be discussed on relevant PRs.

Once the core block examples are sorted my plan is to work on display of the examples in the Color tab.

ramonjd committed:
Split files

ramonjd committed:
Rename cat.name to cat.slug to match categories store object in the block store

update comments

Use `slug` instead of `name`

ramonjd commented on packages/edit-site/src/components/style-book/categories.js:
```diff
@@ -0,0 +1,87 @@
+ * @property {string} slug     Category identifier.
+ * @property {string} title    Category title/label.
+ * @property {Array?} blocks   Array of block names to include in the category. Used when blocks are not included in the category by default.
+ * @property {Array?} excludes Array of blocks to exclude from the category. Used when blocks are included in the category by default.
```
Wondering if I should just convert this all to TS

aaronrobertshaw replied:
One step at a time? 

It could easily be a follow-up but if you're going to pull the trigger on it, maybe before we get too deep into the weeds iterating on then Style Book?

ramonjd replied:
Yeah, good call. That was a 🧠 🌬️

aaronrobertshaw replied:
Didn't see your reply before leaving my review which sort of flip-flopped. I think eventually we'll end up doing it. Now, or in a follow-up, I don't think it matters greatly.

ramonjd replied:
Agree 💯 

Tomorrow problem. 😄

jasmussen commented:
Nice one. Possibly on my end, I got a small error:

<details><summary>Error</summary>


```
TypeError: Cannot read properties of undefined (reading 'examples')
    at http://tt5.local/wp-content/plugins/gutenberg/build/edit-site/index.min.js?ver=8ba71ec71c43f3f0e7b2:33165:35
    at mf (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:105:412)
    at zf (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:127:55)
    at oi (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:125:288)
    at ni (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:124:325)
    at Qk (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:256:479)
    at Ri (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:195:185)
    at Ok (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:195:116)
    at Nd (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:194:492)
    at Qf (http://tt5.local/wp-content/plugins/gutenberg/build/vendors/react-dom.min.js?ver=18:187:439)
```


</details>

What's the best type of feedback I can share at this point? I presume this is initial work to be able to create a new and curated "Overview" page for the book, yes?

ramonjd commented:
> What's the best type of feedback I can share at this point? I presume this is initial work to be able to create a new and curated "Overview" page for the book, yes?

Thank you for testing! 

You're right: this is just the first PR to lay the ground work. The expectations here are that:

- we can control the categories, subcategories and their examples that appear in the style book (achieved via the static list `STYLE_BOOK_CATEGORIES`)
- the categories themselves match, or are close to what the design displays
- the style book works as expected with few or no style changes

This PR will make adding the "Overview" page possible in a follow up PR. 

I'll fix that type error today 🙇🏻

ramonjd commented:
> I'll fix that type error today

✅

aaronrobertshaw reviewed (commented):
Thanks for iterating on this one so quickly @ramonjd, nice work 👍 

I've given this PR a test and it is working well for me.  I did notice one oddity though but it isn't something introduced by this PR.

When in the Style Book, the zoom out button is still available in the top toolbar and has seemingly no effect. If the block inspector button is clicked it takes the user out of the Style Book and back to the normal editor. 

What's the expected behaviour for the Style Book here? I'm not sure it adds much benefit for the Style Book to get zoomed out but others' opinions might vary on that.

As it stands, if the zoom out mode is toggled on in the top toolbar, then the user exits the Style Book, zoom out mode isn't actually in effect. The user then has to toggle the button off, then back on.


https://github.com/user-attachments/assets/7e1ab710-e016-4a85-a974-d93e88315062

---

As for the changes to the Style Book in this PR:

✅ No errors loading style book
✅ Clicking on blocks in the Style Book opens block type under Global Styles sidebar
✅ Current categories are within set outlined by https://github.com/WordPress/gutenberg/issues/53431

> Wondering if I should just cover this all to TS

Thinking on this a little more, maybe it is better to just bite the bullet and convert it to TS now when we're introducing these new files. We can save duplicating the typedefs, keep a cleaner file history etc.

I'm happy to go with your call in the end but I don't think we're in a great rush right?

ramonjd commented:
> Thinking on this a little more, maybe it is better to just bite the bullet and convert it to TS now when we're introducing these new files. We can save duplicating the typedefs, keep a cleaner file history etc.
> 
> I'm happy to go with your call in the end but I don't think we're in a great rush right?

You make a good point about the cleaner file history. I'll see how much work it is. 

I could convert in an immediate follow up so that there are no real functionality changes between the two, and remark on that in the commit. Slightly cleaner than doing a bunch of work in between and then converting. 

🤔 

>  I did notice one oddity though but it isn't something introduced by this PR.
> When in the Style Book, the zoom out button is still available in the top toolbar and has seemingly no effect. If the block inspector button is clicked it takes the user out of the Style Book and back to the normal editor.

Thanks for mentioning. I'll test today and create an issue if one doesn't exist 👍🏻

ramonjd commented:
> Thanks for mentioning. I'll test today and create an issue if one doesn't exist 

https://github.com/WordPress/gutenberg/issues/65630

ramonjd committed:
Create types for utility functions and constants

ramonjd committed:
Fixed regression for styles sidebar: if no category is passed to the Examples component it should render all available examples.
Renamed getCategoryExamples function to getExamplesByCategory

aaronrobertshaw commented on packages/edit-site/src/components/style-book/types.ts:
```diff
@@ -0,0 +1,27 @@
+type Block = {
+	name: string;
+	attributes: Record< string, unknown >;
+	innerBlocks: Block[];
```
Aren't `innerBlocks` optional?

ramonjd replied:
This as well! Good spotting.

aaronrobertshaw commented on packages/edit-site/src/components/style-book/types.ts:
```diff
@@ -0,0 +1,27 @@
+	name: string;
+	title: string;
+	category: string;
+	blocks: Block;
```
Should this be the single Block, an array of blocks e.g. `blocks: Block[];`, or either a single Block or array of Blocks: `blocks: Block | Block[];`?

`getExamples` looks to assign a single block for non-heading blocks, while the example for the heading block is an array of blocks.

The naming of the property indicates an array to me.

FWIW the [draft PR](https://github.com/WordPress/gutenberg/pull/65692) to add a color tab to the Style Book with arbitrary palette color examples [wraps the color blocks in a Grid block](https://github.com/WordPress/gutenberg/pull/65692/files#diff-1b402738fac82aad4bdb2b462dc8e63b59fe80957a7b876d043d9ef110b59a57R82-R96) to provide layout. So it could work with just a single Block.

ramonjd replied:
> Block | Block[];

Oh yes, this! 

Not sure what a better name would be, but I'll fix the type value and mull over it.

aaronrobertshaw reviewed (commented):
Thanks for catching the regression to the top-level style book (when viewed via main site editor nav). The fix tests well for me.

<img width="1343" alt="Screenshot 2024-09-30 at 3 31 20 pm" src="https://github.com/user-attachments/assets/4627910e-2606-466b-a6a9-e5ec03cdc493">

ramonjd committed:
Update types to reflect possible and probable values.
integrate registered categories changes from https://github.com/WordPress/gutenberg/pull/65507

aaronrobertshaw reviewed (approved):
Thanks for the quick fix on the types @ramonjd 🚀 

I've given this another smoke test and it's all testing well for me.

From my end this is looking good to go. If you want to double check the categories with our design gurus, that works for me also. 

🚢

ramonjd commented:
> From my end this is looking good to go. If you want to double check the categories with our design gurus, that works for me also.

Thanks for testing so thoroughly. 

If @jasmussen and folks are okay with the categories in this PR (the only user-facing change). we can move onto the landing page. 👍🏻 

Cheers!

jasmussen commented:
> If @jasmussen and folks are okay with the categories in this PR (the only user-facing change). we can move onto the landing page. 👍🏻

Do you mean these?
![Screenshot 2024-09-30 at 09 24 15](https://github.com/user-attachments/assets/c233dd4f-38bf-4a80-9643-83c6cc229b31)

I suspect as we keep exploring #65619, I suspect the categories will need further iteration regardless, and that work arguably makes the landing page even more important. So no objections on my end, assuming I'm not missing nuance.

ramonjd commented:
Thanks for the quick sanity check! 

 > Do you mean these?

Exactly 👍🏻 

> I suspect the categories will need further iteration regardless

This is the PR that will enable all variations of category lists going forward. The static list can be chopped and changed. I speculate that we'll need to refine its logic as new requirements come it, but right now its _fairly_ flexible. 

Thanks!

matiasbenedetto commented:
The TypeScript code added in this PR seems to be causing errors in code editors like VScode, probably because the `edit-site` does not contain `tsconfig.json` files. It's currently difficult to extend this code because of the numerous errors detected by the TypeScipt engine.


Examples:
![image](https://github.com/user-attachments/assets/b99741e0-2777-4ca2-973c-d77d8f920bb2)

![image](https://github.com/user-attachments/assets/37522e82-8e20-4148-aaed-a103dc261a66)
